pkg/annotation/annotation.go

```diff
@@ -0,0 +1,104 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package annotation
```

Comment:
after reading this PR a number of times, I think I'd prefer a `pkg/lateinitialize` package with a simplified calling interface kind of like this:

(this would be in, say, `pkg/runtime/reconciler.go` [here](https://github.com/aws-controllers-k8s/runtime/blob/a506412923e6667fbc36cc4c841f4c6efe8bd51e/pkg/runtime/reconciler.go#L210))

```go

import (
    acklateinit "github.com/aws-controllers-k8s/runtime/pkg/lateinitialize"
)

// Sync ensures that the supplied AWSResource's backing API resource
// matches the supplied desired state.
//
// It returns a copy of the resource that represents the latest observed state.
func (r *resourceReconciler) Sync(
	ctx context.Context,
	rm acktypes.AWSResourceManager,
	desired acktypes.AWSResource,
) (acktypes.AWSResource, error) {
	var err error
	rlog := ackrtlog.FromContext(ctx)
	exit := rlog.Trace("r.Sync")
	defer exit(err)

	var latest acktypes.AWSResource // the newly created or mutated resource

	isAdopted := IsAdopted(desired)
	rlog.WithValues("is_adopted", isAdopted)

	rlog.Enter("rm.ReadOne")
	latest, err = rm.ReadOne(ctx, desired)
	rlog.Exit("rm.ReadOne", err)

	// Check to see if we attempted to late initialize this resource previously and
	// marked the resource as needing to be late initialized after some delay
	if acklateinit.RequiresInitialization(desired) {
		if latest, err = r.lateInitializeResource(ctx, rm, latest); err != nil {
			return latest, err
		}
	}
	if err != nil {
		if err != ackerr.NotFound {
			return latest, err
		}
		if isAdopted {
			return nil, ackerr.AdoptedResourceNotFound
		}
		if latest, err = r.createResource(ctx, rm, desired); err != nil {
			return latest, err
		}
	} else {
		if latest, err = r.updateResource(ctx, rm, desired, latest); err != nil {
			return latest, err
		}
	}
	// Attempt to late initialize the resource. If there are no fields to
	// late initialize, this operation will be a no-op.
	if latest, err = r.lateInitializeResource(ctx, rm, latest); err != nil {
		return latest, err
	}
	return r.handleRequeues(ctx, latest)
}
```

Reply:
Great Suggestion! I agree to implement it but I have two suggestions for the above code.

a) Perform re-attempt of late initialization before the readOne call inside Sync() method. i.e. Make it the very first action in Sync method. Since lateInitialization makes the readOne call, we do not need to make the extra call.

b) We save the lateInitialized desired into the same object and if more requeues are not required, we continue with the rest of Sync function as it was before.

```go
import (
    acklateinit "github.com/aws-controllers-k8s/runtime/pkg/lateinitialize"
)

// Sync ensures that the supplied AWSResource's backing API resource
// matches the supplied desired state.
//
// It returns a copy of the resource that represents the latest observed state.
func (r *resourceReconciler) Sync(
	ctx context.Context,
	rm acktypes.AWSResourceManager,
	desired acktypes.AWSResource,
) (acktypes.AWSResource, error) {
	var err error
	rlog := ackrtlog.FromContext(ctx)
	exit := rlog.Trace("r.Sync")
	defer exit(err)

	var latest acktypes.AWSResource // the newly created or mutated resource

	isAdopted := IsAdopted(desired)
	rlog.WithValues("is_adopted", isAdopted)

        // Check to see if we attempted to late initialize this resource previously and
	// marked the resource as needing to be late initialized after some delay
	if acklateinit.RequiresInitialization(desired) {
                // This method only updates the late initialized spec fields if they are
                // not provided by the user but added by the AWS service. Which makes
                // it safe to capture the result as desired state of k8s resource.
		if desired, err = r.lateInitializeResource(ctx, rm, desired); err != nil {
			return desired, err
		}
	}

	rlog.Enter("rm.ReadOne")
	latest, err = rm.ReadOne(ctx, desired)
	rlog.Exit("rm.ReadOne", err)

	if err != nil {
		if err != ackerr.NotFound {
			return latest, err
		}
		if isAdopted {
			return nil, ackerr.AdoptedResourceNotFound
		}
		if latest, err = r.createResource(ctx, rm, desired); err != nil {
			return latest, err
		}
	} else {
		if latest, err = r.updateResource(ctx, rm, desired, latest); err != nil {
			return latest, err
		}
	}
	// Attempt to late initialize the resource. If there are no fields to
	// late initialize, this operation will be a no-op.
	if latest, err = r.lateInitializeResource(ctx, rm, latest); err != nil {
		return latest, err
	}
	return r.handleRequeues(ctx, latest)
}
```